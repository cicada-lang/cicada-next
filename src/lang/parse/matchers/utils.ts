import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
