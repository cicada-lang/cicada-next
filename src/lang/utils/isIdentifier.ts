const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
