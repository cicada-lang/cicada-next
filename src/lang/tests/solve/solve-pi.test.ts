import { expect, test } from "vitest"
import { runCode } from "../utils"

test("solve Pi", async () => {
  const output = await runCode(`

solve (A: Type, B: Type) {
  unify (A) -> B = (String) -> String
}

`)

  expect(output).toMatchInlineSnapshot('"{ A: String, B: String }"')
})

test("solve Pi -- nested", async () => {
  const output = await runCode(`

solve (A: Type, B: Type) {
  unify (A, String) -> B = (String, String) -> String
  unify (A, String) -> String = (String, String) -> B
}

`)

  expect(output).toMatchInlineSnapshot('"{ A: String, B: String }"')
})

test("solve Pi -- occur twice", async () => {
  const output = await runCode(`

solve (A: Type, B: Type) {
  unify (A, B) -> B = (String, String) -> String
  unify (A, B) -> String = (String, String) -> B
}

`)

  expect(output).toMatchInlineSnapshot('"{ A: String, B: String }"')
})

test("solve Pi -- deepWalk", async () => {
  const output = await runCode(`

solve (A: Type, B: Type, C: Type) {
  // NOTE To test deepWalk, the order matters here.
  unify C = (A) -> B
  unify A = String
  unify B = String
}

`)

  expect(output).toMatchInlineSnapshot(
    '"{ A: String, B: String, C: (_1: String) -> String }"',
  )
})
