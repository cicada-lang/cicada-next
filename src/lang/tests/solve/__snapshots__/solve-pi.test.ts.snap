// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`solve Pi > 
    "{
      A: String,
      B: String
    }"
   1`] = `
"{
  A: String,
  B: String
}"
`;

exports[`solve Pi -- deepWalk > 
    "{
      A: String,
      B: String,
      C: (String) -> String
    }"
   1`] = `
"{
  A: String,
  B: String,
  C: (String) -> String
}"
`;

exports[`solve Pi -- deepWalk: 
    "{
      A: String,
      B: String,
      C: (String) -> String
    }"
   1`] = `
"{
  A: String,
  B: String,
  C: (String) -> String
}"
`;

exports[`solve Pi -- nested > 
    "{
      A: String,
      B: String
    }"
   1`] = `
"{
  A: String,
  B: String
}"
`;

exports[`solve Pi -- nested: 
    "{
      A: String,
      B: String
    }"
   1`] = `
"{
  A: String,
  B: String
}"
`;

exports[`solve Pi -- occur shadowed by Pi > 
    "{
      X: (X1: Type) -> X1
    }"
   1`] = `
"{
  X: (X1: Type) -> X1
}"
`;

exports[`solve Pi -- occur shadowed by Pi: 
    "{
      X: (X1: Type) -> X1
    }"
   1`] = `
"{
  X: (X1: Type) -> X1
}"
`;

exports[`solve Pi -- occur twice > 
    "{
      A: String,
      B: String
    }"
   1`] = `
"{
  A: String,
  B: String
}"
`;

exports[`solve Pi -- occur twice: 
    "{
      A: String,
      B: String
    }"
   1`] = `
"{
  A: String,
  B: String
}"
`;

exports[`solve Pi: 
    "{
      A: String,
      B: String
    }"
   1`] = `
"{
  A: String,
  B: String
}"
`;

exports[`solve Pi 1`] = `
"{
  A: String,
  B: String
}"
`;

exports[`solve Pi -- nested 1`] = `
"{
  A: String,
  B: String
}"
`;

exports[`solve Pi -- occur twice 1`] = `
"{
  A: String,
  B: String
}"
`;

exports[`solve Pi -- deepWalk 1`] = `
"{
  A: String,
  B: String,
  C: (String) -> String
}"
`;

exports[`solve Pi -- occur shadowed by Pi 1`] = `
"{
  X: (X1: Type) -> X1
}"
`;
